src/server/ns_turn_ratelimit.c

```diff
@@ -0,0 +1,123 @@
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "ns_turn_maps.h"
+#include "ns_turn_ioalib.h"
+#include "ns_turn_ratelimit.h"
+
+/////////////////// rate limit //////////////////////////
+
+ur_addr_map *rate_limit_map = NULL;
+int ratelimit_window_secs = RATELIMIT_DEFAULT_WINDOW_SECS;
+TURN_MUTEX_DECLARE(rate_limit_main_mutex);
+
+void ratelimit_add_node(ioa_addr *address) {
+  // copy address
+  ratelimit_entry *rateLimitEntry = (ratelimit_entry *)malloc(sizeof(ratelimit_entry));
+  TURN_MUTEX_INIT(&(rateLimitEntry->mutex));
+  rateLimitEntry->request_count = 1;
+  rateLimitEntry->last_request_time = time(NULL);
+
+  ur_addr_map_put(rate_limit_map, address, (ur_addr_map_value_type)rateLimitEntry);
+}
+
+int ratelimit_delete_expired(ur_map_value_type value) {
+  turn_time_t current_time = turn_time();
+  ratelimit_entry *rateLimitEntry = (ratelimit_entry*)(void*)(ur_map_value_type)value;
+  return (rateLimitEntry->last_request_time + RATELIMIT_DEFAULT_WINDOW_SECS < current_time);
+}
+
+void ratelimit_init_map() {
+  TURN_MUTEX_INIT(&rate_limit_main_mutex);
+  TURN_MUTEX_LOCK(&rate_limit_main_mutex);
+
+  rate_limit_map = (ur_addr_map*)malloc(sizeof(ur_addr_map));
+  ur_addr_map_init(rate_limit_map);
+  TURN_MUTEX_UNLOCK(&rate_limit_main_mutex);
+}
+
+int ratelimit_is_address_limited(ioa_addr *address, int max_requests, int window_seconds) {
+  /* Housekeeping, prune the map when ADDR_MAP_SIZE is hit and delete expired items */
+  turn_time_t current_time = turn_time();
+
+  if (rate_limit_map == NULL) {
+    ratelimit_init_map();
+  }
+
+  if (ur_addr_map_num_elements(rate_limit_map) >= ADDR_MAP_SIZE) {
+    TURN_MUTEX_LOCK(&rate_limit_main_mutex);
+    /* Set ratelimit_window_secs to grant access to our delete function */
+    ratelimit_window_secs = window_seconds;
+
+    addr_list_foreach_del_condition(rate_limit_map, ratelimit_delete_expired);
+    TURN_MUTEX_UNLOCK(&rate_limit_main_mutex);
+  }
+
+  ur_addr_map_value_type ratelimit_ptr = 0;
+
+  ioa_addr *address_new = (ioa_addr *)malloc(sizeof(ioa_addr));
```

Comment:
Is allocation really needed here?
Why not on stack? It is not stored anyway and used temporarily

Reply:
100% correct, removed malloc()/free()